### ISSUE TITLE
Intermittent Audio Dropouts in v0.3.0 (And I don’t Think It’s Audio Drift)

### ISSUE THREAD

**Issue by leafac** [ExistentialAudio/BlackHole#515]
See, for example, https://youtu.be/-tkDzNuGYuM?t=1936

Only happens from time to time.

It wasn’t an issue in v0.2.10.

For what it’s worth, I’m using 4 instances of BlackHole that I compiled myself using the Wiki instructions.

Here are my Audio MIDI Setup settings:

<img width="1072" alt="Screen Shot 2021-12-29 at 19 46 38" src="https://user-images.githubusercontent.com/586813/147698154-7b540536-12e5-4dd8-b9f9-c464ca98caaa.png">

How can I help figure this out? It only happens from time to time and I don’t know how to reproduce it.

**Comment by devinroth** [comment]
Made some pretty significant changes on v0.3.0.

Do you know what triggers the issue?

Devin Roth

**Comment by leafac** [comment]
Unfortunately, not really—it seems pretty random.

On the stream I linked above issues happened around 32:25 and 41:27, for example. These weren’t even the same instance of BlackHole: the first was my microphone feed, and the seconds was the computer sounds.

At first I thought it could be my computer overloading with something, but in the second case I was just writing in a browser…

Started happening today when I updated from v0.2.10 to v0.3.0, so it’s probably some of the newer changes.

**Comment by devinroth** [comment]
Give me your exact setup. I’ll try to reproduce when I can.

Devin Roth

**Comment by leafac** [comment]
<img width="698" alt="Screen Shot 2021-12-29 at 20 58 59" src="https://user-images.githubusercontent.com/586813/147703031-c29330f1-c957-43f0-89c8-3b105129cf07.png">

<img width="642" alt="Screen Shot 2021-12-29 at 21 00 25" src="https://user-images.githubusercontent.com/586813/147703069-9ce437f6-e828-4da0-b27a-860d0a76bc81.png">

All the BlackHole instances are going through REAPER v6.43 to route and add effects. That setup has been working for months and haven’t changed, so I don’t think it’s that part.

What else do you think could be relevant?

**Comment by leafac** [comment]
I did an experiment. I left REAPER outputting a sine wave through 4 BlackHole instances and recorded it back in REAPER. The experiment ran for an hour during which I wasn’t using the computer. No dropouts. Maybe it does have to do with the load on the machine after all. But that’s awkward, given that the changes in v0.3.0 were supposed to improve the performance…

By the way, here are the changes I made to `BlackHole.h`. It’s a bit more than what’s recommended on the Wiki, but I thought they were reasonable. And I bet it doesn’t have to do with the issue at hand, but just in case…

<img width="851" alt="Screen Shot 2021-12-29 at 22 20 07" src="https://user-images.githubusercontent.com/586813/147707603-9182cae2-ac62-49f9-8d20-b14b147f4dc5.png">

**Comment by devinroth** [comment]
I utilized Apple’s Accelerate framework that supposedly boosts performance. Perhaps I can run more tests.

**Comment by leafac** [comment]
For what it’s worth, everything is back to working as normal after the downgrade to v0.2.10. Maybe Apple’s performance promise doesn’t hold up, but I feel like it wasn’t related to the load on my computer—it appeared to be random.

If you think there are more tests that I can run, please let me know.

**Comment by devinroth** [comment]
Thanks for the update.

It’s like whack a mole. Fix one bug and it breaks something else.

I’ll try to get to the bottom of this. I’ll let you know if you can run some tests for me.

Devin Roth

**Comment by devinroth** [comment]
@leafac could you test this out please. https://github.com/ExistentialAudio/BlackHole/tree/0.3.1 I noticed a bug where I think a minor load could cause a dropout. Hopefully that fixes the issue. Let me know.

**Comment by leafac** [comment]
Thanks. I’ll try it and report back.

**Comment by leafac** [comment]
I tested and it seems that the issue has been fixed! 🙌

I was going to test the `0.3.1` branch but then I noticed that there’s a `0.4.0`, so I just went with that instead. I’m running 4 instances of BlackHole—all with the default settings (except for the `kDriver_Name`, of course). [Here’s an example of a stream done with this setup](https://www.youtube.com/watch?v=aIme0Kf0BfA). I’ve done several streams at this point and everything is working perfectly!

Again: thank you very much for your work on BlackHole!
